A26-grpclb-selection.md

```diff
@@ -0,0 +1,135 @@
+A26: gRPCLB Selection
+----
+* Author(s): Doug Fawley
+* Approver: a11r
+* Status: Draft
+* Implemented in: None
+* Last updated: 2019-10-22
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Modify the method for enabling gRPCLB to not require any special handling in
+core, allowing for stabilization of the gRPCLB protocol and load balancing
+policy.
+
+## Background
+
+The existing mechanism to enable and select gRPCLB is currently still
+experimental. In addition, gRPCLB itself is deprecated, to be replaced by xDS as
+the official advanced load balancing protocol for gRPC ([announcement
+email](https://groups.google.com/d/msg/grpc-io/0yGihF-EFQo/A4QKdXffBwAJ)). We
+would like to enable production uses of gRPCLB, but cannot do so if gRPCLB
+support is not stable. Therefore, we need a mechanism for selecting gRPCLB that
+can be declared stable, yet not create long-lasting maintenance problems.
+
+## Proposal
+
+C core will promote its existing channel arg that configures the DNS resolver to
+fetch SRV records to "stable".
+
+In Go and Java, the default DNS resolver will not fetch SRV records. However,
+when gRPCLB is imported, it will override the default DNS resolver with a
+resolver that does fetch SRV records. Justification for this approach: without
+the gRPCLB LB policy, balancer addresses have no value. Users that do not wish
+to use gRPCLB will not take the performance penalty caused by attempting to
+resolve the balancer addresses through SRV records.
+
+Additionally, in all languages, to remove special handling of gRPCLB from the
+channel:
+- The loadBalancingConfig field in service config will be the only way to
```

Comment:
If I'm correct, this is an intentionally breaking change, that will cause clients that are currently enabling grpclb policy via SRV only (and not resolving or statically configuring <i>any</i> service config), to no longer use grpclb policy.

If that's the case, then don't see issues with this in itself, but at least a handful of experimental users would be broken by this. I.e., AFAICS we need to plan a migration and then help migrate those users, as a part of doing this.

Reply:
It should already be possible to enable grpclb via the service config even if there are no balancer addresses, so I think there's a transition plan.  We change the experimental users' code first and then do a new gRPC release with this change.

Reply:
In Go, since our gRPCLB support is not experimental, we will keep the existing selection logic, probably throughout v1.